### PR TITLE
[AutoDiff] WIP - Adding support for differentiating wrt inout parameters.

### PR DESCRIPTION
**Pull request by eaplatanios**
@rxwei This is a very first step but I thought it'd be good to start making progress here and keep all the related discussions here.

### CODE REVIEW

**Comment by eaplatanios** [comment]
@rxwei in the last commit I added a helper for handling `inout` types when compute the JVP and VJP types. I'll pause now until you take a look and tell me if this looks ok and what the next steps are.

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -4471,6 +4471,17 @@ Optional<VectorSpace> TypeBase::getAutoDiffAssociatedTangentSpace(
   return cache(None);
 }
 
+Type getTangentType(Type type) {
```
Comment (rxwei) [inline lib/AST/Type.cpp]: This should be defined as a closure inside `AnyFunctionType::getAutoDiffAssociatedFunctionType`.

**Comment by eaplatanios** [comment]
@rxwei @dan-zheng I also added support for inout wrt parameters to `AttributeChecker::visitDifferentiableAttr`. It doesn't yet support propagating property `@differentiable` to both the setter and getter, but will add support for that once you confirm that my current approach is correct/reasonable.

**Comment by eaplatanios** [comment]
I actually added support for propagating the `@differentiable` attribute to the getters and setters of properties. I'm going to bed now, but once you verify that this looks good, it'd be good to go through the next steps in a bit more detail because I'm not familiar at all with the SIL optimizer phases. I just noticed that we need to make changes [here](https://github.com/apple/swift/blob/6b822d8ac5ae878278e5f53c669d9917367f029f/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1446) and [here](https://github.com/apple/swift/blob/6b822d8ac5ae878278e5f53c669d9917367f029f/lib/SILOptimizer/Mandatory/Differentiation.cpp#L3144) probably, but not sure what those should be exactly.

**Comment by eaplatanios** [comment]
I also ran tests to see if all is still good and everything passes except for an unexpected error message. Adding `@differentiable` to property setters results in `cannot differentiate void function '_' with no inout wrt parameters`, which is weird because it means that setters do not have inout parameters, or that automatically populating the wrt parameters currently ignores inout parameters.

**Comment by eaplatanios** [comment]
@rxwei I started looking a bit into the SIL optimization phases to try and figure out what to do next, but I had one question about function parameters: what's the difference between direct and indirect parameters? I assume I should use something like `SILParameterInfo::isIndirectInOut()` to detect the inout arguments?

**Comment by rxwei** [comment]
Indirect parameters are parameters that need to be handled as buffers due to various reasons (unknown size, `inout`, etc). The [abstraction difference](https://github.com/apple/swift/blob/master/docs/SIL.rst#abstraction-difference) section in the SIL reference has some explanation. `inout` parameters are always indirect.

**Comment by eaplatanios** [comment]
@rxwei I've going through the code and I'm wondering if we also need to change `TypeBase::getAutoDiffAssociatedTangentSpace` so that it can handle `inout` types. If so, I'm not sure how to change it because I'm not sure whether `inout T` will be interpreted as conforming to `Differentiable` if `T` conforms. I believe we should have to chance the `Differentiable` conformance derivation somewhere to deal with this, so that we can also make sure that the tangent space is `inout`.

**Comment by eaplatanios** [comment]
The reason I'm asking this is because I don't know if [this condition](https://github.com/apple/swift/blob/4ce1aaa2fb441f938ed32d1ad2ed8e797c5be061/lib/SILOptimizer/Mandatory/Differentiation.cpp#L2018) will be true for `inout` types.

**Comment by eaplatanios** [comment]
So currently in this PR I have made a modification in `AnyFunctionType::getAutoDiffAssociatedFunctionType` mapping the inout types like this:
```c++
  auto getTangentType = [lookupConformance](Type type) -> Type {
    if (!type->is<InOutType>()) {
      return type->getAutoDiffAssociatedTangentSpace(
          lookupConformance)->getType();
    }
    Type base = type->getInOutObjectType();
    Type tangentBase = base->getAutoDiffAssociatedTangentSpace(
        lookupConformance)->getType();
    return InOutType::get(tangentBase);
  };
```
Should I instead move this inside `TypeBase::getAutoDiffAssociatedTangentSpace` so it's propagated wherever else that function is used?

**Comment by eaplatanios** [comment]
Actually, I realize I cannot currently do that directly because `VectorSpace` does not support the notion of `inout` so in order to move the mapping inside `TypeBase::getAutoDiffAssociatedTangentSpace` we would need to modify `VectorSpace` as well. @rxwei please let me know which option you think is preferable.

**Comment by rxwei** [comment]
An `inout`-type doesn't formally exist in the language. It's just a compiler data structure to represent parameters that are `inout`. A tangent type by itself should never be `inout`. Instead of adding logic to `TypeBase::getAutoDiffAssociatedTangentSpace` or writing a function like this, we should handle this specifically in the place where we compute tangent/cotangent's parameter types.

**Comment by eaplatanios** [comment]
I see. In that case, I guess my current implementation as part of `AnyFunctionType::getAutoDiffAssociatedFunctionType` should be ok. Would the next step be to change `SILType::isDifferentiable` and `TypeResolver::isDifferentiableType` so they can handle types marked as `inout`?

**Comment by eaplatanios** [comment]
Actually I just pushed a change doing that because I feel like this is indeed necessary.

**Comment by eaplatanios** [comment]
Also, I don't think I need to change `DifferentiableActivityInfo::analyze` after all because the inout arguments will already be marked as useful if they're being differentiated with respect to and so we don't need to add them to the set of output values at this stage. Is this reasoning correct?

**Comment by eaplatanios** [comment]
I also started making changes to the VJP emitter but I'm not at all sure if this is the right direction right now so I'll pause until you can take a look at the current set of changes and summarize the appropriate next steps.

**Comment by eaplatanios** [comment]
Also cc'ing @dan-zheng because I noticed that you have also worked a lot in this part of the codebase.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by dan-zheng** [comment]
I accidentally deleted `tensorflow` branch, which closed this PR. That was not intentional, sorry!

**Comment by dan-zheng** [comment]
Thank you @eaplatanios for starting this effort and discussion! Sorry for letting this PR rot.

 https://github.com/apple/swift/pull/30013 just added `inout` argument differentiation support. The PR description has examples.
https://github.com/apple/swift/pull/29959 has typing rules for `inout` derivative functions.

**Pull request closed without merging**